crates/rolldown/src/types/scan_stage_cache.rs

```diff
@@ -0,0 +1,98 @@
+use arcstr::ArcStr;
+use oxc_index::IndexVec;
+use rolldown_common::GetLocalDbMut;
+use rolldown_utils::rayon::{IntoParallelRefIterator, ParallelIterator};
+use rustc_hash::FxHashMap;
+
+use crate::{
+  module_loader::module_loader::VisitState,
+  stages::scan_stage::{NormalizedScanStageOutput, ScanStageOutput},
+};
+
+#[derive(Default)]
+pub struct ScanStageCache {
+  cache: Option<NormalizedScanStageOutput>,
+  module_id_to_idx: FxHashMap<ArcStr, VisitState>,
+}
+
+impl ScanStageCache {
+  #[inline]
+  pub fn set_cache(&mut self, cache: NormalizedScanStageOutput) {
+    self.cache = Some(cache);
+  }
+
+  pub fn set_module_id_to_idx(&mut self, module_id_to_idx: FxHashMap<ArcStr, VisitState>) {
+    self.module_id_to_idx = module_id_to_idx;
+  }
+
+  pub fn take_module_id_to_idx(&mut self) -> FxHashMap<ArcStr, VisitState> {
+    std::mem::take(&mut self.module_id_to_idx)
+  }
+
+  pub fn merge(&mut self, mut scan_stage_output: ScanStageOutput) {
+    let Some(ref mut cache) = self.cache else {
+      self.cache = Some(scan_stage_output.into());
+      return;
+    };
+    let modules = match scan_stage_output.module_table {
+      rolldown_common::HybridIndexVec::IndexVec(_index_vec) => {
+        unreachable!()
+      }
+      rolldown_common::HybridIndexVec::Map(map) => map,
+    };
+    // TODO: Considering newly Added module
+    //
+    // merge module_table, index_ast_scope, index_ecma_ast
+    for (new_idx, new_module) in modules {
+      // dbg!(&new_idx);
+      // dbg!(&new_module.id());
+      let idx = self.module_id_to_idx[new_module.id_clone()].idx();
+
+      let old_module = std::mem::replace(&mut cache.module_table.modules[idx], new_module);
+      let Some(new_module) = cache.module_table.modules[idx].as_normal_mut() else {
+        continue;
+      };
+      let old_module = old_module.as_normal().unwrap();
+
+      let new_ecma_ast_idx = new_module.ecma_ast_idx.expect("should have ecma_ast_idx");
+
+      let old_ecma_ast_idx = old_module.ecma_ast_idx.expect("should have ecma_ast_idx");
+
+      new_module.ecma_ast_idx = Some(old_ecma_ast_idx);
+      std::mem::swap(
+        &mut cache.index_ecma_ast[old_ecma_ast_idx],
+        &mut scan_stage_output.index_ecma_ast[new_ecma_ast_idx],
+      );
```

Comment:
Here look like should only move the new ast to cached ast, the `swap` also move the cached ast to new.

Reply:
it's alright, since the `scan_stage_output` will be dropped anyway, the new normalizedScanStageOutput is created by the cache, https://github.com/rolldown/rolldown/blob/5f52fd51222be14dfc4f738f64f570b123737d64/crates/rolldown/src/types/scan_stage_cache.rs?plain=1#L75-L97